ta/pkcs11/src/object.c

```diff
@@ -409,3 +409,591 @@ enum pkcs11_rc entry_destroy_object(struct pkcs11_client *client,
+		uint32_t *handles = NULL;
+
+		/*
+		 * If there are no attributes specified, we return
+		 * every object
+		 */
+		if (req_attrs->attrs_count) {
+			rc = token_obj_matches_ref(req_attrs, obj);
+			if (rc == PKCS11_RV_NOT_FOUND)
+				continue;
+			if (rc != PKCS11_CKR_OK)
+				goto out;
+		}
+
+		rc = check_access_attrs_against_token(session, obj->attributes);
+		if (rc)
+			continue;
+
+		/* Object may not yet be published in the session */
+		obj_handle = pkcs11_object2handle(obj, session);
+		if (!obj_handle) {
+			obj_handle = handle_get(&session->object_handle_db,
+						obj);
+			if (!obj_handle) {
+				rc = PKCS11_CKR_DEVICE_MEMORY;
+				goto out;
+			}
+		}
+
+		handles = TEE_Realloc(find_ctx->handles,
+				      (find_ctx->count + 1) * sizeof(*handles));
+		if (!handles) {
+			rc = PKCS11_CKR_DEVICE_MEMORY;
+			goto out;
+		}
+
+		/* Store object handle for later publishing */
+		find_ctx->handles = handles;
+		*(handles + find_ctx->count) = obj_handle;
+		find_ctx->count++;
+	}
+
+	if (rc == PKCS11_RV_NOT_FOUND)
+		rc = PKCS11_CKR_OK;
+
+	/* Save target attributes to search (if needed later) */
+	find_ctx->attributes = req_attrs;
+	req_attrs = NULL;
+	session->find_ctx = find_ctx;
+	find_ctx = NULL;
+	rc = PKCS11_CKR_OK;
+
+out:
+	TEE_Free(req_attrs);
+	TEE_Free(template);
+	release_find_obj_context(session, find_ctx);
+
+	return rc;
+}
+
+enum pkcs11_rc entry_find_objects(struct pkcs11_client *client,
+				  uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = params;
+	TEE_Param *out = params + 2;
+	enum pkcs11_rc rc = PKCS11_CKR_GENERAL_ERROR;
+	struct serialargs ctrlargs = { };
+	struct pkcs11_session *session = NULL;
+	struct pkcs11_find_objects *ctx = NULL;
+	char *out_handles = NULL;
+	size_t out_count = 0;
+	size_t count = 0;
+	size_t idx = 0;
+
+	if (!client || ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	out_count = out->memref.size / sizeof(uint32_t);
+	out_handles = out->memref.buffer;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rc = serialargs_get_session_from_handle(&ctrlargs, client, &session);
+	if (rc)
+		return rc;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	ctx = session->find_ctx;
+
+	/*
+	 * TODO: should we check again if these handles are valid?
+	 */
+	if (!ctx)
+		return PKCS11_CKR_OPERATION_NOT_INITIALIZED;
+
+	for (count = 0, idx = ctx->next; idx < ctx->count; idx++, count++) {
+		struct pkcs11_object *obj = NULL;
+
+		if (count >= out_count)
+			break;
+
+		TEE_MemMove(out_handles + count * sizeof(uint32_t),
+			    ctx->handles + idx, sizeof(uint32_t));
+		ctx->next = idx + 1;
+
+		if (idx < session->find_ctx->temp_start)
+			continue;
+
+		/* Newly published handles: store in session list */
+		obj = handle_lookup(&session->object_handle_db,
```

Comment:
why do we need to do this?

Reply:
Hmmm, you're right, not needed. Thanks!

Reply:
Then I believe there is no need for `temp_start`? And above if comes redundant.

Reply:
`temp_start` is needed by `release_find_obj_context()`. It prevents having entries in the session object database that client does not access. When freeing find context, the handlers already returned shall not be released, only those not fetched by client.
An alternate way could be to keep these objects in the session db until the session is closed, and get rid of `temp_start` as suggested.

In the mean time, here, could be change to:
```c
	for (count = 0, idx = ctx->next; idx < ctx->count, count < out_count;
	     idx++, count++)
		TEE_MemMove(out_handles + count * sizeof(uint32_t),
			    ctx->handles + idx, sizeof(uint32_t));

	ctx->next = idx;
```

Reply:
> Unrelated thought -> handle_put() seems to be quiet mis-leading. Somehow the name doesn't suggest we are trying to destroy a handle. Should we change it to handle_destroy()

handle.c exports `handle_get()` and `handle_put()` to create destroy handles. I think __get/put__ are often used  for such purpose.

> Look at the Testcase: (...) So this testcase assumption that same handle will be available across 2 sessions seems wrong to me. Do you agree ?

Yes, I agree, the test should not succeed.

Reply:
I have fixes for session usage in my tree's. I will give a PR for those soon and also fix some other session related things that are in master now.

Reply:
As I was going through the logic again, there is an issue with the way use of temp_start. In release_find_obj_context(), using temp_start would even erase the handles of the token objects from the session->object_handle_db which were even created by this session making that handle as invalid for that session which is not the right case/
eg.
C_OpenSession()
Create token objects -> save handles
C_FindInit() -> find all token objects.
C_FindFinal()
[In above operation we have skipped the C_Find()]
or a scenario where you try to get all token objects but retrieve only 1 of them using C_Find() and call C_FindFinal().
If we try and use the object handles saved in step 2 above, these would be considered as invalid handles which is not correct.

One way of dealing with this scenario would be to remove the usage of temp_start all together as @etienne-lms suggested.  The only issue with this would be if a token object which was not created by this session was added to database in C_FindInit() but not fetched in C_Find(). We would have an extra object handle with session in this case which application is not aware of. This however would not lead to any error scenario.

So, I am proposing that we remove temp_start all together. @vesajaaskelainen , @etienne-lms , let me know your thoughts.

Reply:
I don't catch the issue. In the scenario
> C_OpenSession()
Create token objects -> save handles
C_FindInit() -> find all token objects. - (
C_FindFinal()

the created object handles were already published in the session, so C_FindInit() lists them before `temp_start`.
If there were other token object not already opened by this session, each gets a handle value (`handle_get()`). C_FindFinal() only destroy handles (`handle_put()`) never output to client: from `find_ctx->next` position, at least starting from `temp_start` position.
So `handle_put()` is never called for a handle previously output to the client.

Reply:
The logic in C_FindInit() iterates through the session object list first. This would not have any token objects created by the session. Then it iterates over the token object list , checks if handle of token object found exists in the object_handle_db else adds it there. temp_start if set to the index of first token object. Thus even published token objects end up after token_start.

I have also added a test case for same in xtest, which fails with current implementation.

Reply:
No, published token objects were already in `session->object_handle_db` since published, so found at 1st pass.
The pass on token object skips those already in `session->object_handle_db` (case they already have a handle).

Thanks for the added test.

Reply:
Despite my arguments above, i think it's fine with your simplification, removing the handle release optimization. I would not expect a client keeps a unique pkcs11 session to seek into the many objects and to perform target crypto operations.

Reply:
What you are saying is true. However, temp_start indicates the index in find_ctx and is set to find_ctx->count which indicates  only the session objects. The pass on token objects skips adding the token objects already in object_handle_db. However they are all added in the find_ctx->handle.

This the reason why the subcase SubTest 7 I moved in the recent test failed when it tried to find a token object after doing a find operation.